Exceptions when features are installed before KAR installation completed

Major issue with snapshot 2441 (first startup), apparently a problem to install rrd4j.
What is strange is that I am using the kar file which normally contains everything and so nothing should require a download. This looks like a "corrupted" kar file.
```
org.apache.karaf.features.internal.util.MultiException: Error:
        Error downloading mvn:org.openhab.addons.bundles/org.openhab.persistence.rrd4j/3.1.0-SNAPSHOT
        at org.apache.karaf.features.internal.download.impl.MavenDownloadManager$MavenDownloader.<init>(MavenDownloadManager.java:91)
        at org.apache.karaf.features.internal.download.impl.MavenDownloadManager.createDownloader(MavenDownloadManager.java:72)
        at org.apache.karaf.features.internal.region.Subsystem.downloadBundles(Subsystem.java:457)
        at org.apache.karaf.features.internal.region.Subsystem.downloadBundles(Subsystem.java:452)
        at org.apache.karaf.features.internal.region.SubsystemResolver.resolve(SubsystemResolver.java:224)
        at org.apache.karaf.features.internal.service.Deployer.deploy(Deployer.java:399)
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.doProvision(FeaturesServiceImpl.java:1069)
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.lambda$doProvisionInThread$13(FeaturesServiceImpl.java:1004)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)
        Suppressed: java.io.IOException: Error downloading mvn:org.openhab.addons.bundles/org.openhab.persistence.rrd4j/3.1.0-SNAPSHOT
                at org.apache.karaf.features.internal.download.impl.AbstractRetryableDownloadTask.run(AbstractRetryableDownloadTask.java:77)
                at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
                at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
                at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
                ... 3 more
        Caused by: java.io.IOException: Error resolving artifact org.openhab.addons.bundles:org.openhab.persistence.rrd4j:jar:3.1.0-SNAPSHOT: [Could not find artifact org.openhab.addons.bundles:org.openhab.persistence.rrd4j:jar:3.1.0-SNAPSHOT]
                at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.configureIOException(AetherBasedResolver.java:803)
                at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.resolve(AetherBasedResolver.java:774)
                at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.resolve(AetherBasedResolver.java:657)
                at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.resolve(AetherBasedResolver.java:598)
                at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.resolve(AetherBasedResolver.java:565)
                at org.apache.karaf.features.internal.download.impl.MavenDownloadTask.download(MavenDownloadTask.java:52)
                at org.apache.karaf.features.internal.download.impl.AbstractRetryableDownloadTask.run(AbstractRetryableDownloadTask.java:60)
                ... 6 more
                Suppressed: shaded.org.eclipse.aether.transfer.ArtifactNotFoundException: Could not find artifact org.openhab.addons.bundles:org.openhab.persistence.rrd4j:jar:3.1.0-SNAPSHOT
                        at shaded.org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:403)
                        at shaded.org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifacts(DefaultArtifactResolver.java:215)
                        at shaded.org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifact(DefaultArtifactResolver.java:192)
                        at shaded.org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveArtifact(DefaultRepositorySystem.java:247)
                        at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.resolve(AetherBasedResolver.java:767)
                        ... 11 more
        Caused by: shaded.org.eclipse.aether.resolution.ArtifactResolutionException: Error resolving artifact org.openhab.addons.bundles:org.openhab.persistence.rrd4j:jar:3.1.0-SNAPSHOT
                at shaded.org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:413)
                at shaded.org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifacts(DefaultArtifactResolver.java:215)
                at shaded.org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifact(DefaultArtifactResolver.java:192)
                at shaded.org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveArtifact(DefaultRepositorySystem.java:247)
                at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.resolve(AetherBasedResolver.java:767)
                ... 11 more
```
@kaikreuzer : certainly something critical for the final 3.1.

--- a/bundles/org.openhab.core.karaf/pom.xml
+++ b/bundles/org.openhab.core.karaf/pom.xml
@@ -56,11 +56,18 @@
       <scope>provided</scope>
     </dependency>
     <dependency>
+      <groupId>org.apache.karaf.kar</groupId>
+      <artifactId>org.apache.karaf.kar.core</artifactId>
+      <version>${karaf.tooling.version}</version>
+      <scope>provided</scope>
+    </dependency>
+    <dependency>
       <groupId>org.openhab.core.bundles</groupId>
       <artifactId>org.openhab.core.io.rest</artifactId>
       <version>${project.version}</version>
       <scope>compile</scope>
-    </dependency>
+    </dependency>>
+
   </dependencies>
 
 </project>

--- a/bundles/org.openhab.core.karaf/pom.xml
+++ b/bundles/org.openhab.core.karaf/pom.xml
@@ -58,7 +58,7 @@
     <dependency>
       <groupId>org.apache.karaf.kar</groupId>
       <artifactId>org.apache.karaf.kar.core</artifactId>
-      <version>${karaf.tooling.version}</version>
+      <version>${karaf.compile.version}</version>
       <scope>provided</scope>
     </dependency>
     <dependency>

--- a/bundles/org.openhab.core.karaf/pom.xml
+++ b/bundles/org.openhab.core.karaf/pom.xml
@@ -66,8 +66,7 @@
       <artifactId>org.openhab.core.io.rest</artifactId>
       <version>${project.version}</version>
       <scope>compile</scope>
-    </dependency>>
-
+    </dependency>
   </dependencies>
 
 </project>

--- a/bundles/org.openhab.core.karaf/src/main/java/org/openhab/core/karaf/internal/FeatureInstaller.java
+++ b/bundles/org.openhab.core.karaf/src/main/java/org/openhab/core/karaf/internal/FeatureInstaller.java
@@ -16,6 +16,7 @@ import static java.util.function.Predicate.not;
 
 import java.io.FileInputStream;
 import java.io.IOException;
+import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.util.ArrayList;
@@ -33,6 +34,7 @@ import java.util.Properties;
 import java.util.Set;
 import java.util.concurrent.Executors;
 import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
 import java.util.concurrent.TimeUnit;
 import java.util.function.Predicate;
 import java.util.stream.Collectors;
@@ -40,6 +42,8 @@ import java.util.stream.Stream;
 
 import org.apache.karaf.features.Feature;
 import org.apache.karaf.features.FeaturesService;
+import org.apache.karaf.kar.KarService;
+import org.eclipse.jdt.annotation.Nullable;
 import org.openhab.core.OpenHAB;
 import org.openhab.core.addon.AddonEventFactory;
 import org.openhab.core.common.NamedThreadFactory;
@@ -101,9 +105,11 @@ public class FeatureInstaller implements ConfigurationListener {
 
     private final ConfigurationAdmin configurationAdmin;
     private final FeaturesService featuresService;
+    private final KarService karService;
     private static EventPublisher eventPublisher;
 
     private ScheduledExecutorService scheduler;
+    private @Nullable ScheduledFuture<?> installJob;
 
     private boolean paxCfgUpdated = true; // a flag used to check whether CM has already successfully updated the pax
                                           // configuration as this must be waited for before trying to add feature repos
@@ -121,9 +127,10 @@ public class FeatureInstaller implements ConfigurationListener {
 
     @Activate
     public FeatureInstaller(final @Reference ConfigurationAdmin configurationAdmin,
-            final @Reference FeaturesService featuresService) {
+            final @Reference FeaturesService featuresService, final @Reference KarService karService) {
         this.configurationAdmin = configurationAdmin;
         this.featuresService = featuresService;
+        this.karService = karService;
     }
 
     @Reference
@@ -197,8 +204,33 @@ public class FeatureInstaller implements ConfigurationListener {
                 // which we will receive as another call to modified
                 return;
             }
-            installAddons(config);
+            // if there is an install job running, we cancel it.
+            ScheduledFuture<?> installJob = this.installJob;
+            if (installJob != null) {
+                logger.trace("Configuration changed. Cancelling install job.");
+                installJob.cancel(false); // do not cancel running job, this might cause problems
+                this.installJob = null;
+            }
+            installAddons(config); // we don't have to wait even if the job is running, because method is synchronized
         });
+    }
+
+    private boolean allKarsInstalled() {
+        try {
+            List<String> karRepos = karService.list();
+            Dictionary<String, Object> felixProperties = configurationAdmin
+                    .getConfiguration("org.apache.felix.fileinstall~deploy").getProperties();
+            String addonsDirectory = (String) felixProperties.get("felix.fileinstall.dir");
+            if (addonsDirectory != null) {
+                return Files.list(Path.of(addonsDirectory)).map(Path::getFileName).map(Path::toString)
+                        .filter(file -> file.endsWith(".kar"))
+                        .map(karFileName -> karFileName.substring(0, karFileName.lastIndexOf(".")))
+                        .allMatch(karRepos::contains);
+            }
+        } catch (Exception ignored) {
+        }
+        logger.warn("Could not determine addons folder, its content or the list of installed repositories!");
+        return false;
     }
 
     private void waitForConfigUpdateEvent() {
@@ -332,7 +364,7 @@ public class FeatureInstaller implements ConfigurationListener {
                     }
                 }
                 if (changed) {
-                    properties.put(PROPERTY_MVN_REPOS, repoCfg.stream().collect(Collectors.joining(",")));
+                    properties.put(PROPERTY_MVN_REPOS, String.join(",", repoCfg));
                     paxCfg.update(properties);
                 }
             } catch (IOException e) {
@@ -344,6 +376,15 @@ public class FeatureInstaller implements ConfigurationListener {
     }
 
     private synchronized void installAddons(final Map<String, Object> config) {
+        if (!allKarsInstalled()) {
+            // some kars are not installed, delay installation for 15s
+            logger.info("Some .kar files are not installed yet. Delaying add-on installation by 15s.");
+            installJob = scheduler.schedule(() -> installAddons(config), 15, TimeUnit.SECONDS);
+            return;
+        }
+
+        installJob = null; // either this is called by the job itself or it is null anyway
+
         final Set<String> currentAddons = new HashSet<>(); // the currently installed ones
         final Set<String> targetAddons = new HashSet<>(); // the target we want to have installed afterwards
         final Set<String> installAddons = new HashSet<>(); // the ones to be installed (the diff)

--- a/bundles/org.openhab.core.karaf/src/main/java/org/openhab/core/karaf/internal/FeatureInstaller.java
+++ b/bundles/org.openhab.core.karaf/src/main/java/org/openhab/core/karaf/internal/FeatureInstaller.java
@@ -207,8 +207,8 @@ public class FeatureInstaller implements ConfigurationListener {
             // if there is an install job running, we cancel it.
             ScheduledFuture<?> installJob = this.installJob;
             if (installJob != null) {
-                logger.trace("Configuration changed. Cancelling install job.");
-                installJob.cancel(false); // do not cancel running job, this might cause problems
+                logger.trace("Configuration changed. Cancelling installation job.");
+                installJob.cancel(false); // do not interrupt running job, this might cause problems
                 this.installJob = null;
             }
             installAddons(config); // we don't have to wait even if the job is running, because method is synchronized

